Upgrade OTel from v0.87.0 to v0.96.0

ptodev opened the pull request:
<!--

CONTRIBUTORS GUIDE: https://github.com/grafana/agent/blob/main/docs/developer/contributing.md#updating-the-changelog

If this is your first PR or you have not contributed in a while, we recommend
taking the time to review the guide. It gives helpful instructions for
contributors around things like how to update the changelog.

-->

#### PR Description

OTel upgrade. I went through the changes in the OTel changelog. As far as I can tell, this PR updates all the relevant Agent bits. I hope I didn't miss anything.

#### Which issue(s) this PR fixes

Fixes #6389

#### Notes to the Reviewer

I tested this locally with the Static and River files according to our [docs](https://github.com/grafana/agent/blob/main/docs/developer/updating-otel.md).

There are some additional things I could do:
* Add tests for the new config arguments.
* There are a few TODOs in the changelog which I need to take a look at.
* When the [fork PR](https://github.com/grafana/opentelemetry-collector/pull/10) is merged, I need to update the go.mod file again.
* Update config converters?

#### PR Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] CHANGELOG.md updated
- [x] Documentation added
- [ ] Tests updated
- [ ] Config converters updated

wildum reviewed (commented):
Thanks for taking care of this! I only have a few comments, nothing major. With this review I only checked the added code. I did not go through all opentelemetry release changes to see if something was missing

wildum commented:
Found a missing change: in extract-field-block.md we say that "The source of the labels or annotations. Allowed values are `pod` and `namespace`." But this change in v0.89.0 adds the value "node": https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/28570

wildum commented:
Scrolled through the rest of the open telemetry contrib release, I did not see any other missing changes, well done 👍 For the future I would recommend having atomic commits for every change to make the review easier

ptodev commented on docs/sources/flow/reference/components/otelcol.exporter.otlp.md:
```diff
@@ -177,6 +177,13 @@ information.
+* `rpc_client_request_size_bytes` (histogram): Measures size of RPC request messages (uncompressed).
+* `rpc_client_requests_per_rpc` (histogram): Measures the number of messages received per RPC. Should be 1 for all non-streaming RPCs.
+* `rpc_client_response_size_bytes` (histogram): Measures size of RPC response messages (uncompressed).
+* `rpc_client_responses_per_rpc` (histogram): Measures the number of messages received per RPC. Should be 1 for all non-streaming RPCs.
```
I ran the Agent locally and saw these metrics. They come from one of the libraries, but I can't remember which one.

ptodev commented:
@wildum Thank you for the review! I addressed some of your comments. I plan to address the other ones in a subsequent PR.

I also took out these lines from the CHANGELOG:
```
  * [otelcol.connector.servicegraph] Additional debug metrics
  <!-- TODO: Document those metrics -->
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29917
  * [otelcol.processor.filter] Additional debug metrics
  <!-- TODO: Document those metrics -->
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30736
  * Scope name for all generated Meter/Tracer funcs now includes full package name
  <!-- TODO: Does this affect the Agent's metrics and logs? -->
https://github.com/open-telemetry/opentelemetry-collector/issues/9494
  * [otelcol.connector.servicegraph] Measure latency in seconds instead of milliseconds
  <!-- TODO: This is a breaking change for Static mode? -->
  <!-- TODO: Is this a breaking change for Flow? -->
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27488
```

I will review them and reintroduce them in a subsequent PR.

wildum commented on internal/component/otelcol/config_tls.go:
```diff
@@ -81,6 +82,7 @@ func (args *TLSSetting) Convert() *otelconfigtls.TLSSetting {
 		MinVersion:               args.MinVersion,
 		MaxVersion:               args.MaxVersion,
 		ReloadInterval:           args.ReloadInterval,
+		CipherSuites:             args.CipherSuites,
```
should we create a new slice instead of copying it like this?

wildum commented on internal/static/traces/config_test.go:
```diff
@@ -1159,6 +1159,8 @@ remote_write:
       timeout: 2s
 `,
+			//TODO(ptodev): Look into why we need to add a "cipher_suites: []" explicitly.
+			// The expected config should unmarshal it to [], but instead it sets it to nil.
```
It seems correct to me that the config would unmarshal it to nil. Else maybe you can explicitly create a list during the convert part (converting from nil to {}string[]). I think that we should avoid adding "cipher_suites: []" explicitly

ptodev replied:
> It seems correct to me that the config would unmarshal it to nil.

I used a debugger in one of the Collector's unit tests, to see if it's `nil` or `[]`. It was `[]`, so I assumed that this is the correct thing to do. I don't know if the unit tests deviates from their prod code path, but I guess it doesn't.

ptodev replied:
I think the way Agent's unit test creates the Collector config isn't the same way the Collector does it. Something is different, but I'm not sure what exactly.

wildum reviewed (commented):
Second review, only two little comments

wildum reviewed (approved):
let's ship it!

wildum commented:
@clayton-cornell could you have a look at the docs when you have time? We are merging the branch because this PR is already very big and we don't want it to be stuck longer. Some of my comments and your future comments will be addressed in a subsequent PR
